lib/src/iterable_extensions.dart

```diff
@@ -591,6 +591,45 @@ extension IterableExtension<T> on Iterable<T> {
       yield slice;
     }
   }
+
+  /// Elements paired with their index.
+  ///
+  /// For example, `['A'].enumerated` returns `[Enumerated('A', 0)]`
+  ///
+  /// Common usage:
+  /// ```dart
+  /// for (var el in ['A', 'B', 'C'].enumerated) {
+  ///   print(el.index); // 0, 1, 2
+  ///   print(el.value); // "A", "B", "C"
+  /// }
+  /// ```
+  Iterable<Enumerated<T>> get enumerated sync* {
+    for (var i = 0; i < length; i++) {
+      yield Enumerated<T>(i, elementAt(i));
+    }
+  }
+}
+
+/// An enumerated element from an iterable
```

Comment:
We're considering making `MapEntry` an inline class over a `Record`

I wonder if we'd prefer this also be a Record, and if so, if it would be worth waiting for that language feature to publish this API.

cc @lrhn

Reply:
I had this same thought, seems like a perfect use case for a Record, and waiting for that to land would make sense to me

Reply:
I'd definitely wait for Dart 3.0 and records and then add (maybe even to the platform libraries):
```dart
extension IterableIndexed<T> on Iterable<T> {
  Iterable<(int, T)> get indexed sync* { 
    var index = 0;
    for (var value in this) yield (index++, value);
  }
}
```
(I've already written it into the platform libraries in one of by branches. No promises that it'll land.)